### ISSUE TITLE
Error while update()

### ISSUE THREAD

**Issue by Mohammad-Alavi** [gazugafan/laravel-temporal#6]
I get following error while trying to update my model:

`SQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: wares.id (SQL: insert into "wares" ("id", "owner_id", "parent_id", "name", "price", "cost", "shape", "color", "sku", "barcode", "created_at", "updated_at", "version", "temporal_start", "temporal_end") values (1, 2, ?, Cafe Belbool, 100000, 50000, square, #ffffff, 1000002, 1542365198745, 2020-04-23 19:28:18, 2020-04-23 19:28:18, 2, 2020-04-23 19:28:18, 2999-01-01 00:00:00))`

I managed to fix this by editing "Temporal" trait and removing the ['id], primary key, before insert:

```
//insert the new revision...
$attributes = $this->attributes;
unset($attributes['id']); // i added this line
$query->insert($attributes);
```

This happens because in the line `$attributes = $this->attributes;` we get all the models attribute including 'id' which is primary key and in this line `$query->insert($attributes);` we are trying to insert a new record with the same `id` hence the error. What am i getting wrong?

I am using phpunit with sqlite.
It's worth mentioning that i managed to make this package work with sqlite and there is only this one little error. Ill make a PR when i tested it thoroughly.

Also another thing in the error above at this part
`values (1, 2, ?, Cafe Belbool, ...`
Why am i getting many 'id's? 1, 2, ?

I just wanted your opinion about this error. Do you think this is because i am using sqlite?

**Comment by gazugafan** [comment]
The primary key should be a composite key of both the id and version columns--not just id. You can have lots of rows with the same id, but each one would need to have a different version. From what I remember, composite primary keys are something SQLite can't do. A quick google makes me think I might be wrong about this, though, or maybe that the feature was added recently? I'd try seeing if you can add a composite primary key of id and version.

**Comment by Mohammad-Alavi** [comment]
> If your table already exists and you wish to add a primary key later, you can not use the ALTER TABLE statement to create a primary key. Instead, you must create a new table with the primary key and copy the data into this new table. 

[Reference](https://www.techonthenet.com/sqlite/primary_keys.php)

It's is getting complicated to add support for sqlite 🤔 
Currently you do alter table and add "version" as additional PK, this works for mysql but sqlite ;/

Btw it seems we can have composite key in sqlite.
Still working on it.

**Comment by Mohammad-Alavi** [comment]
@gazugafan 
What do you think is the best way to get the current database driver?
Currently i do it this way:
```
if (strtolower(DB::connection()->getDriverName()) === 'sqlite') {
	// do things here
}
```

I need it to know what driver we are currently using to run different SQL's accordingly.

**Comment by Mohammad-Alavi** [comment]
I couldn't make it to work easily with **sqlite** but it's easily doable if the user do some manual changes.

### TL;DR

Change
`$table->increments('id');`
to
```
$table->unsignedInteger('id');
$table->unsignedInteger('version')->comment('The version number')->default(0)->after('id');
$table->primary(['id', 'version']);
```

and add
```
Schema::table('table_name', function (Blueprint $table) {
	$table->unsignedInteger('id', true)->change();
}
```
after
`Gazugafan\Temporal\Migration::make_temporal('table_name');`

### Longer explanation:

When using **sqlite**:
Because [this](https://github.com/gazugafan/laravel-temporal/issues/6#issuecomment-618707382) it's really gets complicated if we want to add **"version"** column as a second **PK** due to need for creating a whole new table when we need to add another **PK**. But with some manual changes by the user we can add it easily. User have to add **"version"** column themself as **PK** when creating their migration table. Because i use **mysql** as production DB and **sqlite** for testing purposes, this is the way i did it:

```
public function up()
{
    Schema::create('table_name', function (Blueprint $table) {

        if (DB::connection()->getDriverName() === 'mysql') {
            $table->increments('id');
        }
        if (DB::connection()->getDriverName() === 'sqlite') {
	    // Here we add both columns as primary
	    // Notice we didn't specified 'id' as "increments"
	    // "increments" will make column NOT NULL and PRIMARY and we get an error
	    // when we try to add "version" as PRIMARY (u know, because sqlite ;/)
	    // so we make them primary manually and then after "Laravel-Temporal" did its things
	    // (Reference to end of the code) we make 'id' column AUTO INCREMENT
	    $table->unsignedInteger('id');
	    $table->unsignedInteger('version')->comment('The version number')->default(0)->after('id');
	    $table->primary(['id', 'version']);
	}
	$table->timestamps();
	//$table->softDeletes();
});

//add the columns and keys needed for the temporal features...
Gazugafan\Temporal\Migration::make_temporal('table_name');

if (DB::connection()->getDriverName() === 'sqlite') {
    Schema::table('table_name', function (Blueprint $table) {
	// Here we make 'id' column AUTO INCREMENT
	$table->unsignedInteger('id', true)->change();
        }
    });
}
```
Here is the updated [Migration.php](https://gist.github.com/Mohammad-Alavi/08f97f5e8b3f163e811e8be6dd246d61).
What do you think? I know it's ulgy but how can we make this more steamlined?

**Comment by gazugafan** [comment]
> 
> 
> @gazugafan
> What do you think is the best way to get the current database driver?
> Currently i do it this way:
> 
> ```
> if (strtolower(DB::connection()->getDriverName()) === 'sqlite') {
> 	// do things here
> }
> ```
> 
> I need it to know what driver we are currently using to run different SQL's accordingly.

No idea! That looks reasonable enough to me, though.

**Comment by gazugafan** [comment]
For your other comment... as long as you end up with a composite key of id and version, and id auto-increments, I think you should be good! Does laravel support composite primary keys now? I think that was a missing feature before. It looks like you're doing it all via laravel's DB abstraction layer, so I guess so!

If that's the case, you can probably completely remove the make_temporal method. It was just there for convenience since laravel didn't support composite keys. You needed to run raw SQL to make it happen, so I figured a convenience method to do all that made the most sense. If you can do it normally via laravel's DB methods, though, it's probably better to just do that.

**Comment by Mohammad-Alavi** [comment]
> For your other comment... as long as you end up with a composite key of id and version, and id auto-increments, I think you should be good! Does laravel support composite primary keys now? I think that was a missing feature before. It looks like you're doing it all via laravel's DB abstraction layer, so I guess so!
> 
> If that's the case, you can probably completely remove the make_temporal method. It was just there for convenience since laravel didn't support composite keys. You needed to run raw SQL to make it happen, so I figured a convenience method to do all that made the most sense. If you can do it normally via laravel's DB methods, though, it's probably better to just do that.

No, Laravel still doesn't support composite key. So we still have to do it your way. And for the sqlite support, i couldn't come up with a cleaner way to do it. For now i think my implementation of sqlite support is not good enough for a PR because the complications that come with it but if you like i can make a PR with this changes and update the doc instructions in case someone needed to use this awesome pkg with sqlite.

Btw thank you very very much for your work on laravel-temporal 🥇 This package is really awesome 🥂

**Comment by gazugafan** [comment]
Thanks! Yeah, I think SQLite support still needs to be relegated to an extremely "experimental" status. I'll just link anyone that asks about it to this issue for now :)
